compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -182,9 +298,20 @@ struct DistributeTransferRead final
           rewriter, indices, offsets, vectorLayout, readOp.getPermutationMap(),
           warpIndices, threadIndices);
 
+      VectorValue slicedMask = nullptr;
+      if (mask) {
+        SmallVector<int64_t> sliceMaskOffsets =
+            getDistributedTransferOffsetsFromNestedLayout(rewriter, offsets,
+                                                          vectorLayout);
+        SmallVector<int64_t> strides(innerVectorType.getRank(), 1);
+        slicedMask = rewriter.create<vector::ExtractStridedSliceOp>(
+            readOp.getLoc(), mask, sliceMaskOffsets, innerVectorType.getShape(),
+            strides);
```

Comment:
I think this needs to account for the transfer permutation map because the mask is applied *before* broadcasting/permutations
```
An optional SSA value mask may be specified to mask out elements read from the MemRef/Tensor. The mask type is an i1 vector with a shape that matches how elements are read from the MemRef/Tensor, before any permutation or broadcasting. Elements whose corresponding mask element is 0 are masked out and replaced with padding.
```
https://mlir.llvm.org/docs/Dialects/Vector/#vectortransfer_read-vectortransferreadop

A little confusing, but makes sense after considering how this operation is lowered to masked loads.

Reply:
I think Im going to bail on broadcasts for now ( i ll add code to fail here) as they are not needed immediately because there is no good way of supporting that unless we change how transfer_reads are lowered.

I ll add support for permutations.

Reply:
after having wrote the code, I dont think the *distribution* has to account for permutations or help me understand why it has to..

As for an example:

pre distribution IR:

```mlir
%41 = vector.create_mask %c128, %dyn, %c1 : vector<128x256x1xi1>
%42 = vector.transfer_read %arg0[%c0, %c0, %c0], %cst_6, %41 {in_bounds = [true, true], permutation_map = affine_map<(d0, d1, d2) -> (d2, d1, d0)>} : memref<128x?x1xf16>, vector<1x256x128xf16>
```

So it is already permuted in the pre-distribution domain.
What actually breaks is the vector layout enforcement that happens early on. So when layout is enforced on the create_mask op, we need to permute layout.

Reply:
> I think Im going to bail on broadcasts for now ( i ll add code to fail here)

Understandable, I think this was one of the main reasons I bailed on masks altogether in the first version of this pattern.

> after having written the code, I dont think the distribution has to account for permutations
> yep. that works. I ve added a test now.

Cool if it works sg (although I'm not sure I follow why exactly). The distribution pattern does unrolling that I thought would have to account for the permutation (even if the distribution portion is handled by the layout propagation like you said).

Reply:
> Cool if it works sg (although I'm not sure I follow why exactly). The distribution pattern does unrolling that I thought would have to account for the permutation (even if the distribution portion is handled by the layout propagation like you said).

well you are right there. 
I need to get StaticTileOffsetRange based on the layout of the mask not the result of the read -- if that make sense.

Reply:
> > I think Im going to bail on broadcasts for now ( i ll add code to fail here)
> 
> Understandable, I think this was one of the main reasons I bailed on masks altogether in the first version of this pattern.

With same logic, i ve added this now and it works.
In the vector layout enforcement, we just needed to drop the broadcasted dims.